### PR TITLE
[RNMobile] Prevent "Undo Level" after Setting Featured Image via Image Block

### PR DESCRIPTION
**Pull request by SiobhyB**
This PR is intended as a temporary workaround for https://github.com/WordPress/gutenberg/issues/32937

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3682

## Description

As outlined in https://github.com/WordPress/gutenberg/issues/32937, the undo/redo functionality doesn't work as expected after setting a featured image via the image block.

After setting a featured image via the block and then tapping undo, it appears as though the action has successfully reverted, as the **Featured** banner is removed from the block. However, the image still remains as the post's featured image.

With this PR, the issue is worked around by preventing an "undo level" from being created after a featured image is set via the image block. Tapping **undo** or **redo** immediately after setting a featured image via the image block should no longer have any effect.

Note, this PR doesn't solve the original issue. In fact, it will introduce a new UX issue, since the undo/redo buttons won’t work as the user would expect i.e. an undo will revert changes they made just before setting a featured image, rather than the featured setting, and may be confusing. However, this inconvenience should be considered preferable to the current experience in production, where changes to the featured image via the image block are not saved, despite the UI indicating the undo is successful. 

The **Future Considerations** section of this PR can be referred to for further hopes/considerations for any future iterations.

## How has this been tested?

* In the app, navigate to _Home_ → _Blog Posts_ to create a new blog post, one which doesn't already have a featured image set.
* Add a new image block and upload an image to it.
* Tap the image block's cog/gear icon to access its settings and then tap the **Set as Featured** button at the bottom.
* Verify that you see a confirmation message about the newly set featured image and that the block now has a **Featured** banner overlaying it.
* Tap the **undo** button in the app's toolbar.
* With this PR, tapping undo should have _no effect_ on the post's featured image. Verify that tapping **undo** or **redo** _doesn't_ change the featured image by observing whether there are any changes to the **Featured** banner and verifying that the image remains the same via _Post Settings_.

## Screenshots 

The following screenshots show the before and after footer in the image settings, with the updated footer noting that the undo/redo buttons won't work as expected for users:

| Before  | After |
| ------------- | ------------- |
|  |  |

## Types of changes

This PR introduces a bug fix (a non-breaking change that fixes an issue), with the following main changes to the code:

* [A call to `editEntityRecord` is currently triggered](https://github.com/WordPress/gutenberg/pull/33057/files#diff-d45b169f7a1824d1f08790b97680e012bb30be66ed5714994a3959ba23cdee94R96) whenever a request to change the featured image is received from the apps (i.e. when the `Set/Remove as Featured` button is tapped from the image block). That action creator [includes an `undoIgnore` parameter](https://developer.wordpress.org/block-editor/reference-guides/data/data-core/#editEntityRecord) that makes it possible to ignore an edit in the undo history. For our purposes, that parameter is set to true so that a new undo level *isn't* created after the featured image is updated via the app.
* Additional text has also been added to the bottom of the image block's settings. [The additional text](https://github.com/WordPress/gutenberg/pull/33057/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R568) reads as follows: `Changes to featured image will not be affected by the undo/redo buttons `

## Future Considerations

As a more permanent fix for the original problem, the team discussed adding side effects to the undo/redo buttons, however, this deviates a lot from existing patterns in Gutenberg and is something we decided against pursuing.

The ideal fix would involve completely refactoring the existing data flow so that Gutenberg’s store is the primary location that the apps read from. A refactor would involve the following non-trivial changes:

* Refactoring the _Post Settings_ flow will require changes in the Android and iOS codebases, with careful testing to account for all the unknowns. There’s also the possibility of this raising a wider architectural debate around the _Post Settings_ panel in the apps, potentially including redesigning the whole user flow.
* The logic around a confirmation dialog currently exists on the native side of each app ([here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L820) for Android and [here](https://href.li/?https://github.com/wordpress-mobile/WordPress-iOS/pull/16427/files#diff-beacfffc3f8a87edf94895da980a81e356f1b5aa7998f70791bc7baec9a11731R695) for iOS). This will need to be refactored so that it lives on the JS side. Again, this would be a desirable endpoint that would actually aid in implementing [this enhancement](https://github.com/WordPress/gutenberg/issues/30410). It has also been suggested that the confirmation dialog could be removed completely, which would be an interesting idea to explore with the design team.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by SiobhyB** [comment]
Noting here that this PR isn't a direct fix for the core issue, which is that the post's featured image doesn't update in accordance with the undo/redo actions. Instead, it's more of a workaround where a new "undo level" simply isn't created when an image is set as featured via the image block in the mobile app.

While exploring possible solutions to this issue, I experimented with the following thoughts: 

* Set up a listener in [the image block's `edit.native.js` file](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js), then send back an updated ID each time a change to the post's featured image is detected. This seemed similar to an earlier approach I'd taken to the overall featured functionality in the image block, but this was eventually ruled out for a few reasons, one being that it would create a lot of requests for image-heavy posts.
* Explore ways this is solved in other areas of the app. Nothing stood out to me as being directly relevant to this issue while searching through mobile issues that mentioned "undo" in [the Gutenberg](https://github.com/WordPress/gutenberg/pulls?q=is%3Apr+label%3A%22Mobile+App+-+i.e.+Android+or+iOS%22+undo+is%3Aclosed) and [the Gutenberg Mobile](https://github.com/wordpress-mobile/gutenberg-mobile/pulls?q=is%3Apr+undo+is%3Aclosed+) repos. As the native apps are being used as the "source of truth" when it comes to updating a featured image's ID, approaches like [the one followed for adding undo/redo support for post titles](https://github.com/WordPress/gutenberg/pull/13514/files) won't work for our needs. I couldn't find a solution that didn't seem like it would muddy the way data is flowing.

Although the "ideal" solution would be for the post's featured image to actually change with the undo/redo functionality, I think the workaround in this PR may be the best compromise given the above considerations and the issue's relatively low priority.

@hypest, I wondered if I could get your thoughts on this whenever you have the chance? This isn't a high priority and I wanted to ask you as you're the most familiar with how data is currently flowing for this issue. I'd be really grateful to hear your thoughts and whether I'm possibly missing another approach.

I'll keep this issue as a draft until I'm a bit more certain on the approach to take.

**Comment by hypest** [comment]
> @hypest, I wondered if I could get your thoughts on this whenever you have the chance? This isn't a high priority and I wanted to ask you as you're the most familiar with how data is currently flowing for this issue. I'd be really grateful to hear your thoughts and whether I'm possibly missing another approach.

Aha, interesting problem to have in our hands Siobhan, thanks for pinging. The way I see and understand it, I would explore how perhaps to attach a "side effect" action to undo/redo since, in the current way we have the data flow set up, we expect the native side to update the internal state of the `feature_media` attribute in GB. Which means that currently, we need to call the native side to modify the featured image id. I'm referring to a "side effect" here in the sense that the undo/redo won't change any data inside Gutenberg, but we want to trigger a new "call to native side to change the id", which will then have the added effect of updating the id kept in the `editorStore`. Does that make sense?

Alternatively, but perhaps a bigger refactor could be to change the data flow in such a way that the editor would only depend and change the internal state kept in the `editorStore` but we'll put listeners in place to detect the changes and propagate the side effects to/from the native side. This could potentially unify the editor side support, but sounds like a bigger ask at this point. At this stage, I'd explore the first idea first, to attach side-effect callbacks to undo/redo for calling the native side.

**Comment by SiobhyB** [comment]
Thanks @hypest, that makes sense and sounds inline with what I'd been experimenting with, without success so far. I'll continue experimenting, as it sounds like there should be a way to achieve this and it'd be preferable to preventing the undo level.

**Comment by hypest** [comment]
> and it'd be preferable to preventing the undo level.

Indeed, I feel that suppressing/preventing undo/redo would mainly be perceived as a bug by the user anyway so, maybe not much of a value exchanging one buggy behavior with another. That said, this has the signs of a fix that could spiral out to a whole project, so we might need to reassess our plan after the additional exploration.

**Comment by SiobhyB** [comment]
I've added details to the **Future Considerations** this PR's description to give further context around why we've decided to go this route, following the existing comments here and internal discussion.

@fluiddot, I've requested your review, and welcome anyone else's opinion here. 🙇‍♀️ 

I'd especially love to hear feedback on text I've added to the confirmation notice: `(Note, this action can only be undone within the block\'s settings or the Post Settings panel.)`My aim was to be as succinct as possible, given the small amount of space and time it's displayed on screen. Open to any feedback or suggestions for improvement!

**Comment by fluiddot** [comment]
> I've added details to the **Future Considerations** this PR's description to give further context around why we've decided to go this route, following the existing comments here and internal discussion.
> 
> @fluiddot, I've requested your review, and welcome anyone else's opinion here. 🙇‍♀️
> 
> I'd especially love to hear feedback on text I've added to the confirmation notice: `(Note, this action can only be undone within the block\'s settings or the Post Settings panel.)`My aim was to be as succinct as possible, given the small amount of space and time it's displayed on screen. Open to any feedback or suggestions for improvement!

I think the message is clear but I'm concerned about the small amount of time to read it, I'm wondering if we could display it within the block settings, similar as we do in other cases like in the Columns block, wdyt?

Columns block settings|Image block settings
--|--
|

**Comment by SiobhyB** [comment]
@fluiddot, good idea! I'd only be a little worried about possibly cluttering the settings with details that may not be of importance for the majority. But, I agree it'd be more readable than adding this to the notice.

For the mock of this, I updated the text to the following: `Note: The editor's undo/redo buttons won't revert any changes to the featured image. The setting/removal of featured images can only be done within the block's settings or the Post Settings panel.` Here's how it looks:



I think it works a bit better than the notice changes. Any thoughts on its design or the updated text?

cc-ing @iamthomasbishop for thoughts and the final say on this, too.

**Comment by fluiddot** [comment]
> @fluiddot, good idea! I'd only be a little worried about possibly cluttering the settings with details that may not be of importance for the majority. But, I agree it'd be more readable than adding this to the notice.

Good point, from my POV I think it makes sense to have it here but I agree that may not be of importance for the majority. However, at the same time, if we don't consider it important enough probably we shouldn't even display it in the notice 😄 .

> For the mock of this, I updated the text to the following: `Note: The editor's undo/redo buttons won't revert any changes to the featured image. The setting/removal of featured images can only be done within the block's settings or the Post Settings panel.` Here's how it looks:

Not sure if it's doable but I'd expect the message to be right after the link and before the separator line, otherwise it feels like it's not related to it, wdyt?

Screenshot 2021-07-27 at 17 26 17

> I think it works a bit better than the notice changes. Any thoughts on its design or the updated text?

The text looks great and very clear to me 👍 .

**Comment by SiobhyB** [comment]
Thanks, Carlos! 🙇‍♀️ 

I'm in favour of having the text in the settings given further thought. Although it might not be relevant to _everyone_ accessing the setting, I do think it holds some valuable details that we shouldn't completely leave users to figure out by themselves.

> Not sure if it's doable but I'd expect the message to be right after the link and before the separator line, otherwise it feels like it's not related to it, wdyt?

I kind of like the text after the separator, so that it doesn't completely detract attention away from the button, but see your point of view too and could be swayed in either direction. @iamthomasbishop, do you have any strong opinions here? If not, I'll go with Carlos' preference and update the PR for review.

**Comment by iamthomasbishop** [comment]
@SiobhyB Thanks for the ping! A few things I'd like to suggest, with hopes of finding a balance between clarity and simplicity here.

- First and foremost, I agree that it's a good idea to have some form of messaging on the settings sheet, but imo it's a bit wordy/complex, so perhaps we can make it more concise and simplify the wording. What do you think about something along the lines of: `Changes to featured image will not be affected by undo/redo.` or `Changes to featured image are not accessible through undo/redo.`? The idea here is to shorten it to the essential—I'm not sure how helpful the part in the current message about _where_ to access the setting is as necessary as the undo/redo constraint.
- Similarly, the copy that shows on the Notice upon set/remove is a bit longer than I'd like to use in a Notice; typically I try to keep it to a few words, similar to a Snackbar component. Personally, I think we can probably revert back to the "current" copy, as long as we have the message on the settings bottom sheet. If you feel strongly about keeping a message there, let's try to find a way to shorten it considerably.
- I believe we typically put labels like this below the cell's border, so the placing it below the button border makes sense in this context. One thing I'd like to do here is align the text label to the left — in some cases centered buttons work well, but in this case it has started to throw the alignment/rhythm off. Can we try that?
- Is there any way we can keep the bottomSheet open when the user taps on the set/remove action? I was on the fence about this during the last iteration, but after seeing it for a bit I think it'd make sense to keep the user in place (on the settings bottomSheet).

Here is a quick sketch of what I'm imagining for these changes:

<a href="https://user-images.githubusercontent.com/1200432/127574131-232edb1d-716a-4716-98f5-967083a88faf.png"></a>

Separately, I want to propose a couple of other things now that we've seen this in use for a bit. These can probably be split into other tickets and aren't urgent.

- I'd like to add a section heading for Featured Image, to break up the "link settings" section before it to add some separation, if possible.
- I'm not sure if this one has been discussed already, but it would be useful to be able to clear a featured image from the "edit" menu at the top-right of an Image block. The flow would look something like this (piggy-backing on my proposed copy changes from earlier):

<a href="https://user-images.githubusercontent.com/1200432/127573032-49f1847c-83da-45d6-ab5f-76664d2b174d.png"></a>

Let me know if you have any thoughts or disagree w/ any of the above. Happy to iterate on it, and it'd probably be good to get a copy review from editorial when we get to a good place 😄 .

**Comment by SiobhyB** [comment]
Thank you so much for the help, @iamthomasbishop!

> First and foremost, I agree that it's a good idea to have some form of messaging on the settings sheet, but imo it's a bit wordy/complex, so perhaps we can make it more concise and simplify the wording. What do you think about something along the lines of: Changes to featured image will not be affected by undo/redo. or Changes to featured image are not accessible through undo/redo.? The idea here is to shorten it to the essential—I'm not sure how helpful the part in the current message about where to access the setting is as necessary as the undo/redo constraint.

I'm all for the idea of making the wording more concise, I agree the previous copy is verbose and likely overly cautious on my side. I worry that users won't recognise what we mean by "undo/redo" and wanted to be careful not to give an impression that they're not able to revert changes to their featured image _at all_. 

How about tweaking your wording a little to make it abundantly clear what we mean by "undo/redo"? `Changes to featured image will not be affected by the undo/redo buttons` perhaps?

> it'd probably be good to get a copy review from editorial when we get to a good place

That's a good idea! I've added the `Needs Copy Review` label for @automattic/editorial and also added a `Copy Review Request` section to the PR's description to sum things up. For their notes, here are the ideas we've thrown around so far:

* `Note, this action can only be undone within the block\s settings or the Post Settings panel.`
* `Note: The editor's undo/redo buttons won't revert any changes to the featured image. The setting/removal of featured images can only be done within the block's settings or the Post Settings panel.`
* `Changes to featured image will not be affected by undo/redo.`
* `Changes to featured image are not accessible through undo/redo.`
* `Changes to featured image will not be affected by the undo/redo button.`

I'll address your other feedback in separate comments too.

**Comment by SiobhyB** [comment]
> Similarly, the copy that shows on the Notice upon set/remove is a bit longer than I'd like to use in a Notice; typically I try to keep it to a few words, similar to a Snackbar component. Personally, I think we can probably revert back to the "current" copy, as long as we have the message on the settings bottom sheet. If you feel strongly about keeping a message there, let's try to find a way to shorten it considerably.

That makes total sense! Now that we're moving the note about undo/redo to the settings panel, I'll remove it from the notice.

> I believe we typically put labels like this below the cell's border, so the placing it below the button border makes sense in this context. One thing I'd like to do here is align the text label to the left — in some cases centered buttons work well, but in this case it has started to throw the alignment/rhythm off. Can we try that?

> I'd like to add a section heading for Featured Image, to break up the "link settings" section before it to add some separation, if possible.

I made a first stab at implementing these changes. Does this screenshot match up to what you're imagining here?

**Comment by SiobhyB** [comment]
> Is there any way we can keep the bottomSheet open when the user taps on the set/remove action? I was on the fence about this during the last iteration, but after seeing it for a bit I think it'd make sense to keep the user in place (on the settings bottomSheet).

This should be achievable but not as trivial as it may appear due to this bug on the iOS side: https://github.com/WordPress/gutenberg/issues/29672. That bug means that, by default, the app will crash on iOS if we attempt to display the settings and the confirmation dialog (which pops up when replacing a featured image) at the same time. We could look into the workaround mentioned in the comments of that GitHub issue or consider removing the confirmation dialog altogether.  

I can create a separate GitHub issue for further discussion.

> I'm not sure if this one has been discussed already, but it would be useful to be able to clear a featured image from the "edit" menu at the top-right of an Image block. The flow would look something like this (piggy-backing on my proposed copy changes from earlier):

That'd be a cool iteration, I've created a separate GitHub issue for this at https://github.com/WordPress/gutenberg/issues/33819. :)

**Comment by iamthomasbishop** [comment]
> How about tweaking your wording a little to make it abundantly clear what we mean by "undo/redo"? Changes to featured image will not be affected by the undo/redo buttons perhaps?

This makes sense 👍 I noticed in [this comment](https://github.com/WordPress/gutenberg/pull/33057#issuecomment-890001942) that it's written as singular ("button"), would you mind updating in your comment (second-to-last bullet-point near the end)?

> I made a first stab at implementing these changes. Does this screenshot match up to what you're imagining here?

(Re: the new section heading) Yea, that's what I was imagining. We may want to remove the border between the button cell and footer cell because this is a single-cell section, if that's not a bother. Other than that, I think this looks and reads nicely. Here's an example:

<a href="https://user-images.githubusercontent.com/1200432/127709938-18fd07a6-c53d-443d-bbd6-d95f8d935b2a.png"></a>

Looking at the standard footer message component, I believe all we'd have to do here is remove the padding-top to "collapse", but essentially what I'm going for is for the spacing to be equidistant between the button and its neighbors: 

<a href="https://user-images.githubusercontent.com/1200432/127710576-d47553be-e3ee-4cb7-97cf-55af76a5a6ff.png"></a>

> I can create a separate GitHub issue for further discussion.

> I'll create a separate GitHub issue for this and update this comment when it's created. :)

Thanks for creating tickets for these other ones as well! 🙌

**Comment by clucasrowlands** [comment]
Hi all — jumping in from Editorial... Looking at your shortlisted copy options above, the one I think you've landed on currently seems to be the most clear:

> Changes to featured image will not be affected by the undo/redo buttons.

Here are two other alternatives you could consider, which flips the copy around to avoid passive language (eg: "will not be affected"):

> The undo/redo actions will not affect changes to the featured image.

> Tapping undo/redo arrows will not affect changes to the featured image.

But that said, what you've currently got works fine!

**Comment by iamthomasbishop** [comment]
Wow, that was fast! Thank you @clucasrowlands!

@SiobhyB let's roll with `Changes to featured image will not be affected by the undo/redo buttons.`. 

I think we can also remove the `Note: ` prefix, although doing this will probably make "buttons" an [orphan](https://opusdesign.us/wordcount/typographic-widows-orphans/) at certain viewport widths, but let's try that and see how the text looks when it wraps.

**Comment by SiobhyB** [comment]
Thank you so much @clucasrowlands. ❤️ I've gone ahead to update the text to be `Changes to featured image will not be affected by the undo/redo buttons.` 

> We may want to remove the border between the button cell and footer cell because this is a single-cell section, if that's not a bother.

> Looking at the standard footer message component, I believe all we'd have to do here is remove the padding-top to "collapse", but essentially what I'm going for is for the spacing to be equidistant between the button and its neighbors:

@iamthomasbishop, I went ahead to remove the bottom border from the button in https://github.com/WordPress/gutenberg/pull/33057/commits/89f15ac3194348db92a207ecacae507f569d5252 and also removed the bottom padding in https://github.com/WordPress/gutenberg/pull/33057/commits/6ac5e797187a5aaaf07a938dca8d01292f32c319, with the following results: 

| Before  | After |
| ------------- | ------------- |
|  |  |

Removing the bottom padding actually served to increase the spacing between the top of the button and the section's "Featured Image" header. I believe that's related to flexbox. I was going to dig deeper but then realised the spacing between the section's header and the button now matches the amount of spacing between other section headers and their content e.g. the "Link Settings" header and the "Link To" tab. I wanted to check with you to see if this is a desired end point or if you'd prefer to see some more tweaks around the spacing here? Thanks in advance!

**Comment by iamthomasbishop** [comment]
> and also removed the bottom padding

@SiobhyB Just wanted to clarify, I'd like to keep the button sizing as-is so that we provide an adequate tap-target size. So the heading and footer cells would get collapsed, like this:

<a href="https://user-images.githubusercontent.com/1200432/127926375-4e5acbb7-6a0f-4d7d-895d-82910cb1ee55.png"></a>

Looking closer at your "after" screenshot above, it looks like the spacing between header/button/footer is equal, but in practice it looks like a bit too much space. Would it be possible to remove the spacing here?

<a href="https://user-images.githubusercontent.com/1200432/127928065-23afaadc-686a-4501-8466-203b82782965.png"></a>

I'm assuming that will require a change to the cell components themselves, so if we're not able to do this easily, I'm okay with either of the options in your before/after above — with a slight preference toward the latter.

**Comment by SiobhyB** [comment]
@iamthomasbishop, I've been having a bit of a difficult time trying to style the components in the way set out in the sketch today. 

The padding beneath the `Featured image` header is being set [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/panel/body.native.scss#L5) and I haven't been able to find a way to target it directly from the image block's `styles.native.scss` file for the image block. Amongst other things, I've tried adding a custom class to that section, but no luck.

I also believe the current spacing is a result of flexbox and have been trying to change things, but again no luck in getting close to the sketch.

@fluiddot, do you perhaps see any approaches or options I'm missing here?

**Comment by iamthomasbishop** [comment]
@SiobhyB That all makes sense. If it makes it easier, let's just roll with your "before" example above—meaning the one that has the separator between the button and footer cell, like this:

<a href="https://user-images.githubusercontent.com/1200432/128254134-b3e62321-6f9f-4d53-88a7-31155018558e.png"></a>

I was on the fence, but I think having a separator between is probably "technically" correct from a design perspective so we can roll with that.

**Comment by SiobhyB** [comment]
@iamthomasbishop, thanks for your continued help/feedback here! I've been able to get this closer to the design you set out [here](https://github.com/WordPress/gutenberg/pull/33057#issuecomment-891357909) following a session with @fluiddot:



Let me know if you'd prefer a change in the padding or if you want the border at the top of the footer note, as that should be fairly straightforward to change now.

@fluiddot, I think the PR is ready for your review now, with the note that there may be some small changes to the CSS following any further design feedback. :) 

~There's also an installable build for Android available [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15072/#issuecomment-886647259) and I'll work to create one for iOS now that the iOS side of the functionality has been merged.~ Edited: Installable builds not currently available due to bridge conflicts in `develop`. I'll aim to update when they're resolved.

**Comment by iamthomasbishop** [comment]
That looks great @SiobhyB , thank you!

File: packages/block-library/src/image/styles.native.scss
Comment (fluiddot) [inline packages/block-library/src/image/styles.native.scss]: I'm wondering if for consistency with the rest of the names we should include the word `Button` here, resulting in the style name `.setFeaturedButtonCellContainer`, wdyt?

File: packages/block-library/src/image/edit.native.js
Comment (fluiddot) [inline packages/block-library/src/image/edit.native.js]: I tested these changes but I'm getting an error that states that the function `getSetFeaturedButton` doesn't exist, looks like the function name has been changed and in fact, doesn't exist, however, I see that `getFeaturedButtonPanel` is present. 

I'm wondering if this should be calling `getFeaturedButtonPanel` instead of `getSetFeaturedButton`, wdyt?

**Review by fluiddot — CHANGES_REQUESTED** [review]

File: packages/block-library/src/image/edit.native.js
Comment (SiobhyB) [reply packages/block-library/src/image/edit.native.js]: Ah, yes, thank you for spotting this! The function was updated to `getFeaturedButtonPanel` following feedback on the iOS side, which was merged recently. A merge error on my part is the reason the old name is being referenced here. Fixed in https://github.com/WordPress/gutenberg/pull/33057/commits/658d0054501f589402b674f006caec38aadc893c.

File: packages/block-library/src/image/styles.native.scss
Comment (SiobhyB) [reply packages/block-library/src/image/styles.native.scss]: That makes sense to me, updated in https://github.com/WordPress/gutenberg/pull/33057/commits/2856d5a6ccbbfe3c974991bbd6e986055b2c02e2. :)

File: packages/block-library/src/image/edit.native.js
Comment (fluiddot) [inline packages/block-library/src/image/edit.native.js]: I was wondering if we really need to have the footer message under a new panel body, if the purpose is to prevent rendering a separator, you could disable it by passing the prop `separatorType="none"` to the `BottomSheet.Cell` components defined in the `getFeaturedButtonPanel` function. This way you could render the `FooterMessageControl` component in the same panel body as the featured button.

Comment (SiobhyB) [reply packages/block-library/src/image/edit.native.js]: You're right that there's no meaningful reason for the footer message to exist in its own panel body. I've gone ahead to move it within the featured panel in https://github.com/WordPress/gutenberg/pull/33057/commits/56625086337fe41cd4ad18b4240f56f37c2cf3d1, and have removed the separator as per your suggestion. :)

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 ! Amazing work @SiobhyB 💯 !

_I tested this PR on Samsung Galaxy S20 FE 5G (Android 10) and Simulator iPhone 11 Pro Max (iOS 14.4)._

**NOTE:** When opening the Image block settings, I noticed that the following error was logged in the console:
```
ExceptionsManager.js:180 Warning: Cannot update a component (`Unknown`) while rendering a different component (`BottomSheetTextControl`). To locate the bad setState() call inside `BottomSheetTextControl`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
    in BottomSheetTextControl (at edit.native.js:416)
```
Looks like it's related to [this line](https://github.com/WordPress/gutenberg/blob/64fadb53c15f9717101b2c64692eb2c75fcec5c5/packages/block-library/src/image/edit.native.js#L417) because if I set an empty function on that prop the error stops. I think it's out of the scope of this PR but I wanted to let you know as it might have been introduced in this [PR](https://github.com/WordPress/gutenberg/pull/29670), as per [this commit](https://github.com/WordPress/gutenberg/commit/c508c1b0955bdba8a8e8e06140163a94da8f49a3#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R395-R426).

**Comment by SiobhyB** [comment]
Thank you so much for your help, @fluiddot! I'll address the error you've noted in a separate PR. 🙇‍♀️ 

I'll go ahead to merge this PR and also wanted to sum up the following pending issues that have been discussed in the comments: 

<ul>
<li>A more permanent fix for <a href="https://github.com/WordPress/gutenberg/issues/32937" rel="nofollow ugc">the broken undo/redo functionality</a>, with the <strong>Future Considerations</strong> section from this PR's description offering some direction/additional context.</li>
<li>An iteration to <a href="https://github.com/WordPress/gutenberg/issues/30410" rel="nofollow ugc">the current replace/confirmation dialog</a>. As part of this implementation, it would also be good to find a way to <a href="https://github.com/WordPress/gutenberg/issues/33821" rel="nofollow ugc">keep the image block's settings sheet open</a> after the "Set/Remove as Featured" button is tapped.</li>
<li>The addition of the "Set as Featured" option <a href="https://github.com/WordPress/gutenberg/issues/33819" rel="nofollow ugc">to the image block's "edit" menu</a>.</li>
</ul>

The plan is for me to start diving into different areas of the codebase, so there is no guarantee of when the above items will be prioritized or if I'll be the one to work on them in the future, but they're at least documented and I'd be happy to help in any way I can with any of them. Thanks to all involved on this project so far for the help and patience. :)

**Pull request merged**
